srcs/app/config/analyser/ParserStatelessSet.cpp

```diff
@@ -0,0 +1,303 @@
+#include <parser/ParserStatelessSet.hpp>
+
+Parser::StatelessSet::StatelessSet(Engine *parser, API *config) :
+	config_(config),
+	parser_(parser) {}
+
+bool Parser::StatelessSet::ParserErrorPage_(
+					 const std::vector<std::string> &input,
+					 uint16_t *code, std::string *uri) {
+	char *endptr = NULL;
+	if (input.size() != 2)
+		return EXIT_FAILURE;
+	int64_t result = std::strtol(input[0].c_str(), &endptr, 10);
+	if ((endptr && endptr[0] != '\0') || errno ||
+		!HttpStatusCodes::IsValid(result))
+		return EXIT_FAILURE;
+	*code = result;
+	*uri = input[1];
+	return EXIT_SUCCESS;
+}
+
+bool Parser::StatelessSet::ParseIpAddressPort_(const std::string &input,
+										 std::string *errorThrow,
+										 uint16_t *outPort,
+										 uint32_t *outAddress) {
+	const char * addressStr = input.c_str();
+	std::string addTmp;  // we need this object's lifetime to last for the
+						 // entire function
+	int64_t port = 8080;
+	char *endptr = NULL;
+	if (std::count(input.begin(), input.end(), '.') != 3) {
+		port = std::strtol(addressStr, &endptr, 10);
+		if (*endptr || errno || port < 1 || port > UINT16_MAX) {
+			*errorThrow = "`listen' directive port number invalid";
+			return EXIT_FAILURE;
+		}
+		*outAddress = 0;
+		*outPort = static_cast<uint16_t>(port);
+		return EXIT_SUCCESS;
+	}
+	if (std::count(input.begin(), input.end(), ':') == 1) {
+		addTmp = input.substr(0, input.find(':'));
+		std::string portTmp =
+			input.substr(input.find(':') + 1, input.size());
+		port = std::strtol(portTmp.c_str(), &endptr, 10);
+		if ((endptr && *endptr) || errno || port < 1 || port > UINT16_MAX) {
+			*errorThrow = "`listen' directive port number invalid";
+			return EXIT_FAILURE;
+		}
+		addressStr = addTmp.c_str();
+	}
+	const in_addr_t address = inet_addr(addressStr);
+	if (address == static_cast<in_addr_t>(-1)) {
+		*errorThrow = "`listen' directive IP invalid";
+		return EXIT_FAILURE;
+	}
+	*outAddress = ntohl(address);
+	*outPort = static_cast<uint16_t>(port);
+	return EXIT_SUCCESS;
+}
+
+t_parsing_state Parser::StatelessSet::InitHandler(const StatefulSet &data) {
+	(void) data;
+	return Parser::State::K_EXP_KW;
+}
+
+t_parsing_state Parser::StatelessSet::SemicHandler(const StatefulSet &data) {
+	(void) data;
+	return Parser::State::K_EXP_KW;
+}
+
+t_parsing_state Parser::StatelessSet::SyntaxFailer(const StatefulSet &data) {
+	std::stringstream str;
+#ifdef DBG
+	str << "Raw data: \""<< data.GetRawData() << "\"\n";
+	str << "State type: \""<<  data.GetState() << "\"\n";
+#endif
+	str << data.GetErrorMessage();
+	throw SyntaxError(str.str(), LINE);
+}
+
+t_parsing_state Parser::StatelessSet::ExpKwHandlerClose
+													(const StatefulSet &data) {
+	(void)data;
+	parser_->PopContext();
+	return Parser::State::K_EXIT;
+}
+
+bool Parser::StatelessSet::isKwAllowedInCtx_(t_parsing_state kw,
+										t_parsing_state ctx) {
+	if ((ctx != Parser::State::K_LOCATION && ctx != Parser::State::K_SERVER
+		&& ctx != Parser::State::K_INIT)
```

Comment:
No. Only those three. They are, basically, the scopes we may be in. Init is the global scope, location is the scope we open after a 'location path' sequence, and so on

Reply:
In this case, this condition will always be `true`, and we could remove it.

Reply:
I think it's good practice, while attempting to write extensible, maintainable code, to write functions with every sanity check that requires the function to minimally work embedded in itself. We may decide, at a later stage, to add some more handlers, some more contexts, and we would have to make additions in several functions across several files, not just here. This is the place where they are first checked, and collected.
I don't think there is anything wrong with this.

Reply:
I am also checking in every function of the Parser::API:: function that we are in a context that makes sense for the assignment to the data structure to work. This check is also redundant, since they are all checked, in the first instance, in this very same function that you are reviewing right now, but for the same reason, I don't think it's incorrect to make the check.

Just as you would check for a zero denominator in a function that divides integers taking an integer as a parameter, or as you check for a pointer to be not null before you dereference it. You check because you never know where is the function going to be called from, later on.

Reply:
Ok, good points, and thank you, now I understand this part better.